lib/rate_map.c

```diff
@@ -203,3 +210,95 @@ rate_map_shift_by_mass(rate_map_t *self, double pos, double mass)
     double result_mass = rate_map_position_to_mass(self, pos) + mass;
     return rate_map_mass_to_position(self, result_mass);
 }
+
+#ifdef _Static_assert
+_Static_assert(FLT_RADIX == 2, "Base 2 floating point types required");
+#endif
+
+static int
+fast_search_lookup_init_lookups(
+    fast_search_lookup_t *self, const double *values, size_t n_values)
+{
+    int ret = 0;
+    const double *lower, *strict_upper;
+    uint32_t i;
+    double query = -INFINITY;
+
+    lower = values;
+    for (i = 0; i < self->num_lookups; i++) {
+        while (*lower <= query) {
+            lower++;
+        }
+        self->lookups[i].start = lower;
+        strict_upper = lower;
+        query = ldexp(i, -(self->power_shift));
+        assert(strict_upper < values + n_values);
+        while (strict_upper < values + n_values && *strict_upper <= query) {
+            strict_upper++;
+        }
+        const int64_t num = strict_upper - lower;
+        if (num > INT32_MAX) {
+            ret = MSP_ERR_BAD_PARAM_VALUE; // TODO: consider new err code
+            goto out;
+        }
+        self->lookups[i].num = (size_t) num;
+    }
+out:
+    return ret;
+}
+
+static inline long
+ceil_to_long(double x)
+{
+    return lrint(ceil(x)); // TODO: test use of fast non-portable equivalent
```

Comment:
isn't this the same as ``return (long) ceil(x)``? What's the ``lrint`` for?

Reply:
`lrint` is a misguided attempt at getting something more optimized than a simple cast. My line here is compiling to 14 machine instructions with one call.

Doing `(long) ceil(x)`, not only is much easier to read, but compiles to 2 machine instructions with one call.

It's pretty cool to see all these super fancy new x64_64 instructions. Like one that does the cast in one instruction. I learned x64 long ago and 64-bit is new to me.